lib/pos_staking_rewards.go

```diff
@@ -0,0 +1,282 @@
+
+	// We reward all snapshotted stakes from the current snapshot validator set. This is an O(n) operation
+	// that loops through all of the snapshotted stakes and rewards them one by one.
+	snapshotStakesToReward, err := bav.GetAllSnapshotStakesToReward()
+	if err != nil {
+		return errors.Wrapf(err, "DistributeStakingRewardsToSnapshotStakes: problem retrieving snapshot stakes to reward: ")
+	}
+
+	// If there are no stakes to reward, then there's nothing to be done. Exit early here.
+	if len(snapshotStakesToReward) == 0 {
+		return nil
+	}
+
+	// Loop through all of the snapshot stakes; distribute staking rewards to the staker and commissions to
+	// their validator.
+	for _, snapshotStakeEntry := range snapshotStakesToReward {
+		if snapshotStakeEntry == nil {
+			// This should never happen. If we encounter a nil entry, then the setter for UtxoView.SnapshotStakesToReward
+			// is unexpectedly setting nil values. We just skip such values here.
+			continue
+		}
+
+		// Compute the staker's portion of the staking reward, and the validator's commission.
+		stakerRewardNanos, validatorCommissionNanos, err := bav.computeStakerRewardAndValidatorCommission(
+			snapshotStakeEntry, elapsedFractionOfYear, apy,
+		)
+		if err != nil {
+			return errors.Wrapf(
+				err,
+				"DistributeStakingRewardsToSnapshotStakes: problem computing staker reward and validator commission: ",
+			)
+		}
+
+		// If both the staker reward and the validator commission are zero, then there's nothing to be done.
+		// Move on to the next staker.
+		if stakerRewardNanos == 0 && validatorCommissionNanos == 0 {
+			continue
+		}
+
+		// Reward the staker their portion of the staking reward.
+		if stakerRewardNanos > 0 {
+			if err = bav.distributeStakingReward(snapshotStakeEntry.ValidatorPKID, snapshotStakeEntry.StakerPKID, stakerRewardNanos); err != nil {
+				return errors.Wrapf(err, "DistributeStakingRewardsToSnapshotStakes: problem distributing staker reward: ")
+			}
+		}
+
+		// Reward the validator their commission from the staking reward.
+		if validatorCommissionNanos > 0 {
+			if err = bav.distributeValidatorCommission(snapshotStakeEntry.ValidatorPKID, validatorCommissionNanos); err != nil {
+				return errors.Wrapf(err, "DistributeStakingRewardsToSnapshotStakes: problem distributing validator commission reward: ")
+			}
+		}
+	}
+
+	return nil
+}
+
+func (bav *UtxoView) computeStakerRewardAndValidatorCommission(
+	snapshotStakeEntry *StakeEntry,
+	elapsedFractionOfYear *big.Float,
+	apy *big.Float,
+) (
+	_stakerRewardNanos uint64,
+	_validatorCommissionNanos uint64,
+	_err error,
+) {
+	// Compute the staker's reward amount using a big float math, and immediately convert it to big int
+	// so we can do the remainder of the math using integer operations. This is the only operation where
+	// we need float math.
+	stakerRewardNanos := convertBigFloatToBigInt(
+		computeStakingReward(snapshotStakeEntry.StakeAmountNanos, elapsedFractionOfYear, apy),
+	)
```

Comment:
No action on this, but I am a bit sad that DESO is still in nanos rather than a 1e18 uint256 thing. It means that sufficiently small stakes with a sufficiently small epoch window will round down to zero. This being said, I checked, and with a 1bp APY, which is the lowest we can do above zero right now, and with our epoch time being an hour, you'd still get 11 nanos for a single DESO per epoch. So hopefully we can continue limping along with this 1e9 situation.

Reply:
Not the end of the world here. In a practical case where the APY is ex: 5%, the hourly reward on 1 DESO would be 5707.78 nanos which rounds down to 5707. And 5707/5707.78 = 1bp. So the 1e9 precision doesn't result in a meaningful loss for the small staker.

That said, I totally agree that we're OK to limp along for now, but we are starting to push the boundary on 1e9 scaling becoming less practical.